### PR TITLE
Secondary

### PR DESCRIPTION
**Pull request by ShoiraTa**
# Project 2: Enumerable Methods

In this project, we wrote our own custom Ruby enumerable methods.

1. my_each
2. my_each_with_index
3. my_select
4. my_all?
5. my_any?
6. my_count
7. my_none?
8. my_inject
9. my_map

We ran test cases for every method we created.

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+    i = param1 ? 0 : 1
+    if block_given?
+      (i...array.length).my_each { |el| accum = yield(accum, array[el]) }
+    elsif param1 && param2
+      (i...array.length).my_each { |el| accum.send(param2, array[el]) }
+    elsif param1
+      accum = array[0]
+      (1...array.length).my_each { |el| accum.send(param1, array[el]) }
+    else
+      raise LocalJump
```
Comment (sinansevgi) [inline enumerable.rb]: - [x] You should write the exact name of the error to raise it.
```suggestion
      raise LocalJumpError
```

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    for element in self
+      yield element
+    end
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    array = to_a
```
Comment (sinansevgi) [inline enumerable.rb:13]: - [x] When you call each_with_index on ranges if a block is given it should return the range itself. 
Your my_each_with_index returns the range as converted to an array.

`(5..10).my_each_with_index{|num| num*2}` this expression should return `(5..10)` but your implementation returns `[5, 6, 7, 8, 9, 10]`

You can check the type of parameter before converting it to an array.

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+  end
+
+  def my_count(param = nil)
+    count = 0
+    if block_given?
+      my_each { |i| count += 1 if yield i }
+    elsif param
+      my_each { |i| count += 1 if match?(i, param) }
+    else
+      return length
```
Comment (sinansevgi) [inline enumerable.rb]: - [x] #my_count method fails when I call it on a range.

It should return the number of items in the range
`(1..4).my_count` should return `4`
 `length` method doesn't work with ranges.

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    accum = param1 || array[0]
+    i = param1 ? 0 : 1
+    if block_given?
+      (i...array.length).my_each { |el| accum = yield(accum, array[el]) }
+    elsif param1 && param2
+      (i...array.length).my_each { |el| accum.send(param2, array[el]) }
+    elsif param1
+      accum = array[0]
+      (1...array.length).my_each { |el| accum.send(param1, array[el]) }
+    else
```
Comment (sinansevgi) [inline enumerable.rb]: - [x]  When you pass a symbol as an argument to #my_inject method, It should combine each element of the collection by applying the symbol as a named method.

 without initial value:
 `(5..10).my_inject(:+)` this expression should return `45`
 with initial value:
 `(5..10).my_inject(2, :*)` this expression should return `302400`

**Review by sinansevgi — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️ 
----

Hello, team! 
You have done a great job. 
The methods are passing almost all of the test cases. 🧿 

Some of your methods are missing some behavior that is present in the original Enumerable methods. 

----
## Please check the inline comments for expected behaviors.

----
You can always contact me on [Slack](https://microverse-students.slack.com/team/U016UBBG4GZ) if you need assistance.

**Comment by ShoiraTa** [comment]
Hello, 

We looked through all the comments and fixed the mentioned errors,

Thank you

**Review by pacyL2K19 — APPROVED** [review]
# Approved :medal_military: 
Hello **Team** you made a great job so far, your readme file looks professional and all the tests pass congratulations :tada: 
Now feel free to merge this branch and submit the project 
Go ahead and keep it high 
Happy coding :desktop_computer:

**Pull request merged**